### PR TITLE
Base refactoring with docs to some t/base/*.t files

### PR DESCRIPTION
**Pull request by kozachkov**
Added `use strict; use warnings` to some t/base/*.t files and comment descriptions about their test plans.

### CODE REVIEW

**Comment by jkeenan** [comment]
On behalf of perl5-porters, I would like to thank you for a first-time contribution to the Perl 5 core distribution.

Unfortunately, we won't be able to accept this pull request in its current form.  Here's why.

The test suite is organized into various directories with certain directories -- including `t/base/` -- being reserved for very fundamental concepts that don't depend on our being able to load modules via `use` or `require`.  That means that files in `t/base/` may not contain any `use strict;` or `use warnings;` statements.  See `perldoc perlhack`.  (As a side effect, this means that `git blame` attributes many statements in these test files to Larry Wall's initial commits to Perl 1 back in 1987.)

If perl had strictures or warnings active *by default* (as was proposed for [one version/vision of Perl 7](https://github.com/Perl/perl5/wiki/Defaults-for-v7)), then it's quite possible that the files you have worked on might have to be re-worked to compile with strictures and run warnings-free.  But that's not the current perl's expectation.

We do, however, look forward to your future contributions.

Thank you very much.
Jim Keenan

**Comment by kozachkov** [comment]
@jkeenan Thank you for the response. I removed unnecessary `use strict; use warnings` from these files.

Is the PR is acceptable now in the form of two different commits?

**Comment by Leont** [comment]
> Is the PR is acceptable now in the form of two different commits?

No, it is not. You're changing a bunch of things without good reason, and adding comments that aren't necessarily good value. I would strongly recommend against touching this part of the test suite unless you truly understand what it's trying to test.

**Comment by hvds** [comment]
I'll echo Jim's thanks for your efforts to improve perl here. Based on a quick look, I'm sure there are parts that we would like. But there are quite a few types of changes mixed in here, and it would be easier to review if they were separated out as distinct commits.

The various comments you have added look mostly helpful, and much of the reformatting looks good too (though I'm not personally a fan of the extra spaces in parens you appear to like). I'm less sure about your choice to replace various `if (cond) { ... } else { ... }` constructs with `cond ? ... : ...` - the latter are in principle harder to parse, so in `t/base` when we are still checking whether the absolute basics of the language work correctly we prefer to avoid them (except of course when that is what we are trying to test).

File: t/base/if.t [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
+
 print "1..2\n";
 
-# first test to see if we can run the tests.
+# First test to see if we can run the tests.
 
 $x = 'test';
-if ($x eq $x) { print "ok 1 - if eq\n"; } else { print "not ok 1 - if eq\n";}
-if ($x ne $x) { print "not ok 2 - if ne\n"; } else { print "ok 2 - if ne\n";}
+if ($x eq $x) { print "ok 1 - if eq\n"; } else { print "not ok 1 - if eq\n"; }
```
Comment (karenetheridge) [inline t/base/if.t:10]: these changes add unwanted trailing whitespace.

File: t/base/pat.t [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
 print "1..2\n";
 
-# first test to see if we can run the tests.
+# First test to see if we can run the tests.
 
 $_ = 'test';
-if (/^test/) { print "ok 1 - match regex\n"; } else { print "not ok 1 - match regex\n";}
-if (/^foo/) { print "not ok 2 - match regex\n"; } else { print "ok 2 - match regex\n";}
+/^test/
+  ? ( print "ok 1 - match regex\n" )
```
Comment (karenetheridge) [inline t/base/pat.t:11]: the `?:` operators should be reserved for strict functional expressions -- that is, they should not be used solely to achieve side effects.

File: t/base/rs.t [hunk 3]
```diff
@@ -1,312 +1,455 @@
...
 $teststring2 = "1234567890123456789012345678901234567890";
 
-# Create our test datafile
-1 while unlink 'foo';                # in case junk left around
+# ------------------------------------------------------
+# Remove previous test files if exist
+# ------------------------------------------------------
+# In case of existing files with the same 'foo' name
+# on some systems with versioning file system (e.g VMS).
+# Read more at pod/perport.pod "System Interaction"
```
Comment (karenetheridge) [inline t/base/rs.t]: typo in filename... and I'm not sure if the explanation in this comment is correct.

File: t/base/rs.t [hunk 4]
```diff
@@ -1,312 +1,455 @@
...
-  close TESTFILE;
-  1 while unlink qw(foo.bar foo.com foo.fdl);
-} else {
-  # Nobody else does this at the moment (well, maybe OS/390, but they can
-  # put their own tests in) so we just punt
-  foreach $test ($test_count..$test_count + 3) {
-      print "ok $test # skipped on non-VMS system\n";
-      $test_count++;
-  }
+if ( $^O eq 'VMS' ) {
```
Comment (karenetheridge) [inline t/base/rs.t:95]: reformatting large parts of a file for whitespace-only changes means that 'git blame' now reports the wrong commit, which causes more work for bisection and other archaeological endeavours.

**Review by karenetheridge — CHANGES_REQUESTED** [review]
I do not see that there is sufficient value gained, in proportion to the volume of changes.

File: t/base/if.t (hunk 1 above)
Comment (kozachkov) [reply t/base/if.t:10]: What about symmetry ?

File: t/base/pat.t (hunk 2 above)
Comment (kozachkov) [reply t/base/pat.t:11]: I see. I will return the form.

File: t/base/rs.t (hunk 3 above)
Comment (kozachkov) [reply t/base/rs.t]: The typo, yes, thank you. What explanation should be used here instead?

File: t/base/rs.t (hunk 4 above)
Comment (kozachkov) [reply t/base/rs.t:95]: In perlhack.pod there are such lines:

> Use spaces for indentation, not tab characters.
> 
> The codebase is a mixture of tabs and spaces for indentation, and we
> are moving to spaces only.  Converting lines you're patching from 8-wide
> tabs to spaces will help this migration.

That means that potentially there are a lot of **similar places** in perl core where should be done such reformatting "which cases more work for bisection and other archaeological endeavours". It cannot be avoided it we want to have a modern (not slovenly) perl code in the repo.

**Comment by kozachkov** [comment]
> > Is the PR is acceptable now in the form of two different commits?
> 
> No, it is not. You're changing a bunch of things without good reason, and adding comments that aren't necessarily good value. I would strongly recommend against touching this part of the test suite unless you truly understand what it's trying to test.

What is a good reason for you? For me it is:

- a clean indentation and appearance at whole;
- a way to simplify reading test suits just by overview their description especially for newbies, we do want to extend of potentially perl developers count, doesn't it?
- an adding a documentation notes about what is going on and why. In perlhack.pod there are such parts:

> While it's not necessary for documentation changes, new tests or
> trivial patches, it's often worth explaining how your change works.
> Even if it's clear to you today, it may not be clear to a porter next
> month or next year.

"it may not be clear to a porter next month or next year." -- and there are such places there which are "not be clear" and I added a few comments about how it works.

Code and documentation must be improving -- that is "a good reason" for me.

>  I would strongly recommend against touching this part of the test suite unless you truly understand what it's trying to test.

And what if I "truly understand"?

As regards "against touching this part". Well, there are you who can point to mistakes of such changes and there is a chance that some changes may be useful for Perl Community, doesn't it?

**Comment by kozachkov** [comment]
@hvds "But there are quite a few types of changes mixed in here, and it would be easier to review if they were separated out as distinct commits." you right about "mixed changes". I will try to do it more separate in the future.

**Comment by Leont** [comment]
> a clean indentation and appearance at whole;

_Clean indentation and appearance_ is highly subjective. If we all update the source code to what we think is clean, our git history is going to look like a wikipedia edit war. In general, changing indentation of a piece of code that you have no ownership over is bad form IMHO.

Likewise, adding your personal convention of `=====` and `-----` headers when none of these files did so before seems rather inappropriate. It's generally better to stay in the style of the existing code/documentation.

"Talk first, shoot second" is generally speaking a good principle.

> a way to simplify reading test suits just by overview their description especially for newbies, we do want to extend of potentially perl developers count, doesn't it?

I agree with that in theory, I don't agree with this implementation. I don't think these overviews clarify anything.

> an adding a documentation notes about what is going on and why. In perlhack.pod there are such parts:

Prepending «Create test 'foo' file» before a simple open/print/close sequence does not achieve any of that. It doesn't answer any question a competent perl programmer would have about that bit of code.

It's not that I think that the answers you provide are incorrect, it's that they answer the wrong questions; questions that are useless because the answer can easily be deduced from basic reading of the code. Good documentation answers questions that can't be answered that way (typically these are why questions).

> And what if I "truly understand"?

> As regards "against touching this part". Well, there are you who can point to mistakes of such changes and there is a chance that some changes may be useful for Perl Community, doesn't it?

In general, I don't think an outsider is the right person to answer the questions, I do think they can be the right person to ask questions (e.g. "why isn't any of this using strict or warnings" would have been a great question).

**Comment by rwp0** [comment]
### suggestion

> Tests in the t/base/ directory ought to be runnable with plain miniperl.That is, they should not require Config.pm nor should they require anyextensions to have been built. TEST will abort if any tests in thet/base/ directory fail.

clarify further in [README](https://www.github.com/Perl/perl5/tree/blead/t%2FREADME) linking to `perlhack(1)` (on `strict` / `warnings`) or create (perhaps by subdividing the main one) a new one specific to the subdirectory (ie. `t/base`)

**Comment by Leont** [comment]
> or create (perhaps by subdividing the main one) a new one specific to the subdirectory (ie. t/base)

Yeah, I think that would be very useful.

**Comment by kozachkov** [comment]
> > a clean indentation and appearance at whole;
> 
> _Clean indentation and appearance_ is highly subjective. If we all update the source code to what we think is clean, our git history is going to look like a wikipedia edit war. In general, changing indentation of a piece of code that you have no ownership over is bad form IMHO.
> 
> Likewise, adding your personal convention of `=====` and `-----` headers when none of these files did so before seems rather inappropriate. It's generally better to stay in the style of the existing code/documentation.
> 
> "Talk first, shoot second" is generally speaking a good principle.
> 
> > a way to simplify reading test suits just by overview their description especially for newbies, we do want to extend of potentially perl developers count, doesn't it?
> 
> I agree with that in theory, I don't agree with this implementation. I don't think these overviews clarify anything.
> 
> > an adding a documentation notes about what is going on and why. In perlhack.pod there are such parts:
> 
> Prepending «Create test 'foo' file» before a simple open/print/close sequence does not achieve any of that. It doesn't answer any question a competent perl programmer would have about that bit of code.
> 
> It's not that I think that the answers you provide are incorrect, it's that they answer the wrong questions; questions that are useless because the answer can easily be deduced from basic reading of the code. Good documentation answers questions that can't be answered that way (typically these are why questions).
> 
> > And what if I "truly understand"?
> 
> > As regards "against touching this part". Well, there are you who can point to mistakes of such changes and there is a chance that some changes may be useful for Perl Community, doesn't it?
> 
> In general, I don't think an outsider is the right person to answer the questions, I do think they can be the right person to ask questions (e.g. "why isn't any of this using strict or warnings" would have been a great question).

@Leont

I agree that:

- "Clean indentation and appearance is highly subjective." because of that there are pod/perlstyle.pod and pod/perlhack.pod with rules for such situations.
- "If we all update the source code to what we think is clean, our git history is going to look like a wikipedia edit war." that doesn't refer situations which are explained in aforesaid pod files.
- "Likewise, adding your personal convention of ===== and ----- headers when none of these files did so before seems rather inappropriate. It's generally better to stay in the style of the existing code/documentation." You are absolutely right, that's my fault and I removed it.

I disagree that:

- "In general, changing indentation of a piece of code that you have no ownership over is bad form IMHO." in open-source projects with many contributors must be common rules for common practice for it and in perl repo there are such ones in pod/.
- "I agree with that in theory, I don't agree with this implementation. I don't think these overviews clarify anything." you are right about "I don't think" cause you are already know these parts of test code.
- "Prepending «Create test 'foo' file» before a simple open/print/close sequence does not achieve any of that. It doesn't answer any question a competent perl programmer would have about that bit of code." If you look more carefully into the code which was before my changes you will see that this line there are there already. And if we follow it, that we should removed these line from the tests.
- "It's not that I think that the answers you provide are incorrect, it's that they answer the wrong questions; questions that are useless because the answer can easily be deduced from basic reading of the code" it is a "highly subjective" statement because in most cases there is no necessity to write documentation comments as an experienced perl programmer can understand parts of code just by reading them, doesn't it?
- "In general, I don't think an outsider is the right person to answer the questions, I do think they can be the right person to ask questions (e.g. "why isn't any of this using strict or warnings" would have been a great question)." and another "great question" I guess should be "why in this test code it is used indentation with 2 spaces instead of 4 as it's advised in pod/perlstyle.pod?

File: t/base/pat.t (hunk 2 above)
Comment (kozachkov) [reply t/base/pat.t:11]: I found the same using `?:` for tests t/base/num.t however.

**Comment by jkeenan** [comment]
This p.r. has not found good support from core committers.  I propose that we close it in 7 days unless such support emerges.

**Comment by jkeenan** [comment]
> This p.r. has not found good support from core committers. I propose that we close it in 7 days unless such support emerges.

Closing as per schedule.

**Pull request closed without merging**